File: src/algorithms/searching/binary-search/solution1.go
package main

// O(log(n)) time | O(log(n)) space
func BinarySearch(array []int, target int) int {
	return helper(array, target, 0, len(array)-1)
}

func helper(array []int, target, left, right int) int {
	if left > right {
		return -1
	}
	middle := (left + right) / 2
	potentialMatch := array[middle]
	if target == potentialMatch {
		return middle
	} else if target < potentialMatch {
		return helper(array, target, left, middle-1)
	}
	return helper(array, target, middle+1, right)
}

File: src/algorithms/searching/binary-search/solution1.js
// O(log(n)) time | O(log(n)) space
function binarySearch(array, target) {
	return binarySearchHelper(array, target, 0, array.length - 1);
};

function binarySearchHelper(array, target, left, right) {
	if (left > right) return -1;
	const middle = Math.floor((left + right) / 2);
	const potentialMatch = array[middle];
	if (target === potentialMatch) {
		return middle;
	} else if (target < potentialMatch) {
		return binarySearchHelper(array, target, left, middle -1);
	} else {
		return binarySearchHelper(array, target, middle + 1, right);
	};
};

exports.binarySearch = binarySearch;